TestCases/strayalien/0072-feel-in/0072-feel-in.dmn

```diff
@@ -0,0 +1,2012 @@
+            <text>4 in (2..4]</text>
+        </literalExpression>
+    </decision>
+
+    <decision name="number_008_e" id="_number_008_e">
+        <variable name="number_008_e"/>
+        <literalExpression>
+            <text>5 in (2..4]</text>
+        </literalExpression>
+    </decision>
+
+    <decision name="number_009_a" id="_number_009_a">
+        <variable name="number_009_a"/>
+        <literalExpression>
+            <text>1 in [2..4)</text>
+        </literalExpression>
+    </decision>
+
+    <decision name="number_009_b" id="_number_009_b">
+        <variable name="number_009_b"/>
+        <literalExpression>
+            <text>2 in [2..4)</text>
+        </literalExpression>
+    </decision>
+
+    <decision name="number_009_c" id="_number_009_c">
+        <variable name="number_009_c"/>
+        <literalExpression>
+            <text>3 in [2..4)</text>
+        </literalExpression>
+    </decision>
+
+    <decision name="number_009_d" id="_number_009_d">
+        <variable name="number_009_d"/>
+        <literalExpression>
+            <text>4 in [2..4)</text>
+        </literalExpression>
+    </decision>
+
+    <decision name="number_009_e" id="_number_009_e">
+        <variable name="number_009_e"/>
+        <literalExpression>
+            <text>5 in [2..4)</text>
+        </literalExpression>
+    </decision>
+
+    <decision name="number_010_a" id="_number_010_a">
+        <variable name="number_010_a"/>
+        <literalExpression>
+            <text>1 in [2..4]</text>
+        </literalExpression>
+    </decision>
+
+    <decision name="number_010_b" id="_number_010_b">
+        <variable name="number_010_b"/>
+        <literalExpression>
+            <text>2 in [2..4]</text>
+        </literalExpression>
+    </decision>
+
+    <decision name="number_010_c" id="_number_010_c">
+        <variable name="number_010_c"/>
+        <literalExpression>
+            <text>3 in [2..4]</text>
+        </literalExpression>
+    </decision>
+
+    <decision name="number_010_d" id="_number_010_d">
+        <variable name="number_010_d"/>
+        <literalExpression>
+            <text>4 in [2..4]</text>
+        </literalExpression>
+    </decision>
+
+    <decision name="number_010_e" id="_number_010_e">
+        <variable name="number_010_e"/>
+        <literalExpression>
+            <text>5 in [2..4]</text>
+        </literalExpression>
+    </decision>
+
+    <decision name="number_011" id="_number_011">
+        <variable name="number_011"/>
+        <literalExpression>
+            <text>1 in 1</text>
+        </literalExpression>
+    </decision>
+
+    <decision name="number_011_a" id="_number_011_a">
+        <variable name="number_011_a"/>
+        <literalExpression>
+            <text>1 in 2</text>
+        </literalExpression>
+    </decision>
+
+    <decision name="number_013" id="_number_013">
+        <variable name="number_013"/>
+        <literalExpression>
+            <text>10 in ? = 10</text>
```

Comment:
This model contains a lot of FEEL expression allegedly (if I understand correctly) referencing a variable called `?`, which however is not in scope as none of these Decisions have an informationRequirement referencing a DRGElement named `?`. If my understanding is correct, this implies this DMN model invalid as containing semantic errors.

Reply:
Thanks @tarilabs.  The '?' is specified in (v1.2) grammar rule 51.c (page 131) as a 'special variable' for use with 'in' tests.  The the expression is a 'general' expression  then it may access the being-tested-for value using '?' - so, as far as I understand, it doesn't have to be an info requirement.  It's a bit like the 'item' special var for iterating.  We're supposed to magically put it in there.

Though, I am not sure whether that syntax is applicable outside of DTs.  But, the tests assert that it is.  I thought best to do more than less.

Reply:
I don't believe we are on the same page and risk overlapping different things. I try summarize my understanding below in the hope this helps :)

First of all, my understanding is that `?` is a valid variable name by the spec, regadless of special use cases. 

When `?` is actually used in a Decision Table it might receive a special semantic according to this clause of the DMNv1.2 specification under chapter "8.3.3 Decision Rule metamodel" :

> d) One of the expressions in the inputEntry is a boolean expressions using the special ‘?’ variable and that expression evaluates to true when the inputExpression value is assigned to ‘?’

When you are saying:

>The '?' is specified in (v1.2) grammar rule 51.c (page 131)

I believe what you are referring to here, is Table 50. Table 50 is described in the spec as:

> Note that Grammar Rules (clause 10.3.1.2) are used in decision table condition cells. These decision table syntaxes are defined in Table 50.

I am understanding the spec here is making reference that Table 50 applies to decision table syntax only. 

In conclusion my understanding is:
- if you use `?` in a FEEL expression in a Decision Table inputEntry (cell) and that FEEL expression is a Predicate (returns a boolean type), the semantic of the DMN specification implies `?` value is the value of the relevant Decision Table inputExpression (column)
- if you use `?` in a FEEL expression outside of a Decision Table inputEntry (cell) this ought to be defined as a variable in scope, hence in a DMN model it must be an informationRequirement.

I am happy to be shown where my understanding is wrong! ;)

Reply:
Thanks @tarilabs.  Appreciated.  That makes sense.  I really wasn't too sure whether that applied outside of DTs (hence it is actually my first note above re assumptions)

> I have made some small assumptions here and there. Like:
* The usage of '?' says for decision tables but I'm not sure if only there. I have assumed usage everywhere and asserted upon it for each type.

I am happy to remove those from the tests - I do appreciate the feedback - I'm kind of working a little blind here sometimes .  

Do have have an opinion on whether the _rest_ of the 'in' syntax should be available outside a DT? Like 'between' and the rest of them.

Many thanks,

Greg.

Reply:
You can indeed use grammar rule 51 outside Decision Table.
For instance you can write a FEEL expression as `1 in [1, 2, 3]` is a valid FEEL expression to be used as a DMN Literal expression. It would be a Predicate. It would evaluate to true.

The way I interpret clause of the DMNv1.2 specification under chapter "8.3.3 Decision Rule metamodel"  is something like ~ _"normally in a DT cell you have a FEEL unary test. You could use instead a FEEL expression as long as you have a `?` reference in it and the FEEL expression is a Predicate"_

I am happy to be shown where my understanding is/might be wrong! ;)

Reply:
@tarilabs okay - great.  Those tests took a lot of effort.  Happy to keep those in.  :-)  I'll comment out the '?' tests (pending any further input here) before removing them. 

Thanks again Matteo.  It sure is helpful to get some guidance.

Reply:
Absolutely, please don't rely uniquely on my feedback, this is also an opportunity for a shared understanding of the spec, indeed a goal of the TCK.

Meanwhile it would be helpful you place/move these test either under the CL2 or CL3 directory (and checking for the numbering it takes the next available as commented by others earlier). I would opt at this point for CL3 if that is also the target of your implementation.